blame/blame.go

```diff
@@ -0,0 +1,223 @@
+	// internal blame struct.
+	var err error
+	b.revs, err = revlist.New(b.repo, b.fRev, b.path)
+	if err != nil {
+		return nil, err
+	}
+	sort.Sort(b.revs) // for forward blame, we need the history sorted by commit date
+
+	// allocate space for the data in all the revisions of the file
+	b.data = make([]string, len(b.revs))
+
+	// init the graph
+	b.graph = make([][]vertex, len(b.revs))
+
+	// for all every revision of the file, starting with the first
+	// one...
+	var found bool
+	for i, rev := range b.revs {
+		// get the contents of the file
+		b.data[i], found = git.Data(b.path, rev)
+		if !found {
+			continue
+		}
+		// count its lines
+		nLines := git.CountLines(b.data[i])
+		// create a node for each line
+		b.graph[i] = make([]vertex, nLines)
+		// assign a commit to each node
+		// if this is the first revision, then the node is assigned to
+		// this first commit.
+		if i == 0 {
+			for j := 0; j < nLines; j++ {
+				b.graph[i][j] = vertex(b.revs[i])
+			}
+		} else {
+			// if this is not the first commit, then assign to the old
+			// commit or to the new one, depending on what the diff
+			// says.
+			b.assignOrigin(i, i-1)
+		}
+	}
+
+	// fill in the output results: copy the nodes of the last revision
+	// into the result.
+	fVs := b.graph[len(b.graph)-1]
+	result := make([]*git.Commit, 0, len(fVs))
+	for _, v := range fVs {
+		c := git.Commit(*v)
+		result = append(result, &c)
+	}
+	return result, nil
+}
+
+// this struct is internally used by the blame function to hold its
+// intputs, outputs and state.
+type blame struct {
+	repo  *git.Repository // the repo holding the history of the file to blame
+	path  string          // the path of the file to blame
+	fRev  *git.Commit     // the commit of the final revision of the file to blame
+	revs  revlist.Revs    // the chain of revisions affecting the the file to blame
+	data  []string        // the contents on the file in all the revisions TODO: not all data is needed, only the current rev and the prev
+	graph [][]vertex      // the graph of the lines in the file across all the revisions TODO: not all vertexes are needed, only the current rev and the prev
+}
+
+type vertex *git.Commit // a vertex only needs to store the original commit it came from
+
+// Assigns origin to vertexes in current (c) rev from data in its previous (p)
+// revision
+func (b *blame) assignOrigin(c, p int) {
+	// assign origin based on diff info
+	hunks := diff.Do(b.data[p], b.data[c])
+	sl := -1 // source line
+	dl := -1 // destination line
+	for h := range hunks {
+		hLines := git.CountLines(hunks[h].Text)
+		for hl := 0; hl < hLines; hl++ {
+			// fmt.Printf("file=%q, rev=%d, r=%d, h=%d, hunk=%v, hunkLine=%d\n", file, rev, r, h, hunks[h], hl)
+			switch {
+			case hunks[h].Type == 0:
+				sl++
+				dl++
+				b.graph[c][dl] = b.graph[p][sl]
+			case hunks[h].Type == 1:
+				dl++
+				b.graph[c][dl] = vertex(b.revs[c])
+			case hunks[h].Type == -1:
+				sl++
+			default:
```

Comment:
Branch exhaustiveness, this is automatically handled by latest compiler versions (since 1.4 I think..) It shouldn't apply here because there's no return inside each case.